In [25]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [26]:
table = pd.read_csv("C:\\Users\\wills\\Desktop\\ML\\bank_data.csv")
table

id  age    income  children  gender      region married  car  \
0    ID12101   48  17546.00         1  FEMALE  INNER_CITY      NO   NO   
1    ID12102   40  30085.10         3    MALE        TOWN     YES  YES   
2    ID12103   51  16575.40         0  FEMALE  INNER_CITY     YES  YES   
3    ID12104   23  20375.40         3  FEMALE        TOWN     YES   NO   
4    ID12105   57  50576.30         0  FEMALE       RURAL     YES   NO   
..       ...  ...       ...       ...     ...         ...     ...  ...   
595  ID12696   61  47025.00         2  FEMALE  INNER_CITY      NO  YES   
596  ID12697   30   9672.25         0  FEMALE  INNER_CITY     YES  YES   
597  ID12698   31  15976.30         0  FEMALE        TOWN     YES  YES   
598  ID12699   29  14711.80         0    MALE  INNER_CITY     YES   NO   
599  ID12700   38  26671.60         0    MALE        TOWN      NO  YES   

    savings_acct current_acct mortgage  pep  
0             NO           NO       NO  YES  
1             NO          YES      YES   NO  
2            YES          YES       NO   NO  
3             NO          YES       NO   NO  
4            YES           NO       NO   NO  
..           ...          ...      ...  ...  
595          YES          YES      YES   NO  
596          YES          YES       NO   NO  
597          YES           NO       NO  YES  
598          YES           NO      YES   NO  
599           NO          YES      YES  YES  

[600 rows x 12 columns]

## <h5> 1. Load and preprocess the data using Numpy or Pandas and the preprocessing functions from scikit-learn. Specifically, you need to separate the target attribute ("pep") from the portion of the data to be used for training and testing. You will need to convert the selected dataset into the Standard Spreadsheet format (scikit-learn functions generally assume that all attributes are in numeric form). Finally, you need to split the transformed data into training and test sets (using 80%-20% randomized split). [Review Ipython Notebook examples from Week 4 for different ways to perform these tasks.] </h5>

In [27]:
target_labels = table.pep
table = table.drop(labels='pep', axis=1)
table = table.drop(labels='id', axis=1)
print(target_labels, table)

0      YES
1       NO
2       NO
3       NO
4       NO
      ... 
595     NO
596     NO
597    YES
598     NO
599    YES
Name: pep, Length: 600, dtype: object      age    income  children  gender      region married  car savings_acct  \
0     48  17546.00         1  FEMALE  INNER_CITY      NO   NO           NO   
1     40  30085.10         3    MALE        TOWN     YES  YES           NO   
2     51  16575.40         0  FEMALE  INNER_CITY     YES  YES          YES   
3     23  20375.40         3  FEMALE        TOWN     YES   NO           NO   
4     57  50576.30         0  FEMALE       RURAL     YES   NO          YES   
..   ...       ...       ...     ...         ...     ...  ...          ...   
595   61  47025.00         2  FEMALE  INNER_CITY      NO  YES          YES   
596   30   9672.25         0  FEMALE  INNER_CITY     YES  YES          YES   
597   31  15976.30         0  FEMALE        TOWN     YES  YES          YES   
598   29  14711.80         0    MALE  INNER_CITY     YES   NO

In [28]:
table_dummies = pd.get_dummies(table, columns = ['gender', 'region', 'married', 'car', 'savings_acct', 'current_acct','mortgage'])
table_dummies

age    income  children  gender_FEMALE  gender_MALE  region_INNER_CITY  \
0     48  17546.00         1              1            0                  1   
1     40  30085.10         3              0            1                  0   
2     51  16575.40         0              1            0                  1   
3     23  20375.40         3              1            0                  0   
4     57  50576.30         0              1            0                  0   
..   ...       ...       ...            ...          ...                ...   
595   61  47025.00         2              1            0                  1   
596   30   9672.25         0              1            0                  1   
597   31  15976.30         0              1            0                  0   
598   29  14711.80         0              0            1                  1   
599   38  26671.60         0              0            1                  0   

     region_RURAL  region_SUBURBAN  region_TOWN  married_NO  married_YES  \
0               0                0            0           1            0   
1               0                0            1           0            1   
2               0                0            0           0            1   
3               0                0            1           0            1   
4               1                0            0           0            1   
..            ...              ...          ...         ...          ...   
595             0                0            0           1            0   
596             0                0            0           0            1   
597             0                0            1           0            1   
598             0                0            0           0            1   
599             0                0            1           1            0   

     car_NO  car_YES  savings_acct_NO  savings_acct_YES  current_acct_NO  \
0         1        0                1                 0                1   
1         0        1                1                 0                0   
2         0        1                0                 1                0   
3         1        0                1                 0                0   
4         1        0                0                 1                1   
..      ...      ...              ...               ...              ...   
595       0        1                0                 1                0   
596       0        1                0                 1                0   
597       0        1                0                 1                1   
598       1        0                0                 1                1   
599       0        1                1                 0                0   

     current_acct_YES  mortgage_NO  mortgage_YES  
0                   0            1             0  
1                   1            0             1  
2                   1            1             0  
3                   1            1             0  
4                   0            1             0  
..                ...          ...           ...  
595                 1            0             1  
596                 1            1             0  
597                 0            1             0  
598                 0            0             1  
599                 1            0             1  

[600 rows x 19 columns]

In [29]:
from sklearn.model_selection import train_test_split
bank_train, bank_test, bank_target_train, bank_target_test = train_test_split(table_dummies, target_labels, test_size=0.2, random_state=33)

print (bank_test.shape)
bank_test[0:5]

(120, 19)


age   income  children  gender_FEMALE  gender_MALE  region_INNER_CITY  \
456   32  13267.6         0              1            0                  0   
366   59  29866.9         1              1            0                  1   
131   61  41609.5         3              0            1                  0   
448   53  48971.6         3              0            1                  0   
337   65  38080.9         1              0            1                  1   

     region_RURAL  region_SUBURBAN  region_TOWN  married_NO  married_YES  \
456             0                0            1           0            1   
366             0                0            0           1            0   
131             1                0            0           0            1   
448             1                0            0           0            1   
337             0                0            0           0            1   

     car_NO  car_YES  savings_acct_NO  savings_acct_YES  current_acct_NO  \
456       0        1                0                 1                0   
366       0        1                1                 0                0   
131       1        0                0                 1                0   
448       0        1                0                 1                1   
337       0        1                0                 1                1   

     current_acct_YES  mortgage_NO  mortgage_YES  
456                 1            0             1  
366                 1            1             0  
131                 1            0             1  
448                 0            1             0  
337                 0            0             1

In [30]:
from sklearn import preprocessing
min_max_scaler = preprocessing.MinMaxScaler().fit(bank_train)
bank_train_norm = min_max_scaler.transform(bank_train)
bank_test_norm = min_max_scaler.transform(bank_test)

In [31]:
np.set_printoptions(precision=2, linewidth=80, suppress=True)
bank_train_norm[0:5]
bank_train_norm.shape

(480, 19)

## <h5> 2. Run scikit-learn's KNN classifier on the test set. Note: in the case of KNN, you should first normalize the data so that all attributes are in the same scale (normalize so that the values are between 0 and 1). Generate the confusion matrix (visualize it using Matplotlib), as well as the classification report. Also, compute the average accuracy score. Experiment with different values of K and the weight parameter (i.e., with or without distance weighting) for KNN to see if you can improve accuracy (you do not need to provide the details of all of your experimentation, but provide a short discussion on what parameters worked best as well as your final results). </h5>

In [32]:
from sklearn import neighbors, tree, naive_bayes

n_neighbors = 2

knn_classifier = neighbors.KNeighborsClassifier(n_neighbors)
knn_classifier.fit(bank_train_norm, bank_target_train)

knnpreds_test = knn_classifier.predict(bank_test_norm)
print(knnpreds_test)

['NO' 'YES' 'NO' 'NO' 'YES' 'YES' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'YES'
 'NO' 'YES' 'YES' 'YES' 'NO' 'NO' 'YES' 'NO' 'NO' 'YES' 'NO' 'YES' 'NO' 'NO'
 'YES' 'NO' 'NO' 'NO' 'YES' 'NO' 'YES' 'NO' 'YES' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO'
 'YES' 'NO' 'YES' 'NO' 'NO' 'YES' 'NO' 'YES' 'NO' 'NO' 'YES' 'YES' 'NO' 'YES'
 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'YES' 'NO' 'NO' 'YES'
 'NO' 'NO' 'NO' 'NO' 'YES' 'YES' 'NO' 'YES' 'YES' 'NO' 'YES' 'NO' 'NO' 'NO'
 'NO' 'NO' 'YES' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'YES' 'NO' 'NO' 'NO' 'NO' 'NO'
 'YES' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO' 'YES' 'YES' 'NO' 'NO' 'NO' 'NO' 'NO' 'NO'
 'NO' 'NO' 'NO']


In [33]:
from sklearn.metrics import confusion_matrix, classification_report

knn_confusion_matrix = confusion_matrix(bank_target_test, knnpreds_test)

print(knn_confusion_matrix)
print(classification_report(bank_target_test, knnpreds_test))
print (knn_classifier.score(bank_test_norm, bank_target_test))

[[54 12]
 [33 21]]
              precision    recall  f1-score   support

          NO       0.62      0.82      0.71        66
         YES       0.64      0.39      0.48        54

    accuracy                           0.62       120
   macro avg       0.63      0.60      0.59       120
weighted avg       0.63      0.62      0.61       120

0.625


## <h5> Overall, including distance weighting made accuracy not as high as without it. The best combination of parameters appears to be a K of 10 and no distance weighting with an accuracy of .69. When distance weighting is included, best accuracies tend to cluster around k=5 and without distance weighting it clusters around k=10. Without distance weighting, lower Ks appear to be less accurate then the same Ks with it. </h5>

## <h5> 3. Repeat the classification using scikit-learn's decision tree classifier (using the default parameters) and the Naive Bayes (Gaussian) classifier. As above, generate the confusion matrix, classification report, and average accuracy scores for each classifier. For each model, compare the average accuracy scores on the test and the training data sets. What does the comparison tell you in terms of bias-variance trade-off? </h5>
<h5> 4. Discuss your observations based on the above experiments. </h5>

In [36]:
# creating tree classifier
tree_classifier = tree.DecisionTreeClassifier()
tree_classifier.fit(bank_train, bank_target_train)

treepreds_test = tree_classifier.predict(bank_test)
print(treepreds_test)

#creating Naive Bayes classifier
nb_classifier = naive_bayes.GaussianNB()
nb_classifier = nb_classifier.fit(bank_train, bank_target_train)
nbpreds_test = nb_classifier.predict(bank_test)
print (nbpreds_test)

['NO' 'YES' 'NO' 'YES' 'YES' 'YES' 'YES' 'YES' 'YES' 'NO' 'NO' 'YES' 'NO' 'NO'
 'NO' 'NO' 'NO' 'NO' 'YES' 'NO' 'YES' 'YES' 'NO' 'NO' 'YES' 'NO' 'YES' 'YES'
 'NO' 'YES' 'NO' 'NO' 'NO' 'YES' 'NO' 'YES' 'NO' 'YES' 'NO' 'YES' 'YES' 'NO'
 'NO' 'NO' 'YES' 'NO' 'NO' 'YES' 'NO' 'YES' 'YES' 'YES' 'YES' 'YES' 'NO' 'YES'
 'YES' 'YES' 'NO' 'NO' 'YES' 'NO' 'YES' 'YES' 'YES' 'YES' 'NO' 'NO' 'YES' 'NO'
 'NO' 'NO' 'NO' 'YES' 'NO' 'NO' 'NO' 'NO' 'YES' 'NO' 'NO' 'YES' 'YES' 'NO' 'NO'
 'YES' 'NO' 'NO' 'YES' 'YES' 'NO' 'NO' 'NO' 'NO' 'YES' 'NO' 'NO' 'NO' 'NO' 'NO'
 'NO' 'NO' 'YES' 'YES' 'NO' 'YES' 'YES' 'NO' 'YES' 'NO' 'YES' 'NO' 'NO' 'NO'
 'NO' 'NO' 'NO' 'NO' 'NO' 'NO']
['NO' 'YES' 'NO' 'YES' 'YES' 'NO' 'NO' 'NO' 'NO' 'YES' 'NO' 'YES' 'NO' 'NO'
 'NO' 'NO' 'YES' 'NO' 'YES' 'NO' 'YES' 'YES' 'NO' 'NO' 'NO' 'NO' 'YES' 'YES'
 'NO' 'YES' 'YES' 'NO' 'NO' 'NO' 'NO' 'NO' 'YES' 'YES' 'NO' 'NO' 'NO' 'NO'
 'YES' 'NO' 'YES' 'NO' 'NO' 'NO' 'NO' 'NO' 'YES' 'YES' 'NO' 'NO' 'YES' 'YES'
 'YES' 'NO' 'NO' 'NO' 'YES' 'NO' 'Y

In [38]:
tree_confusion_matrix = confusion_matrix(bank_target_test, treepreds_test)

print(tree_confusion_matrix)
print(classification_report(bank_target_test, treepreds_test))

tree_accuracy_test = tree_classifier.score(bank_test, bank_target_test)
tree_accuracy_train = tree_classifier.score(bank_train, bank_target_train)

nb_confusion_matrix = confusion_matrix(bank_target_test, nbpreds_test)

print(nb_confusion_matrix)
print(classification_report(bank_target_test, nbpreds_test))

nb_accuracy_test = nb_classifier.score(bank_test, bank_target_test)
nb_accuracy_train = nb_classifier.score(bank_train, bank_target_train)

print(f"Tree test accuracy: {tree_accuracy_test}, Tree train accuracy: {tree_accuracy_train}, NB test accuracy: {nb_accuracy_test}, NB train accuracy: {nb_accuracy_train}")


[[56 10]
 [14 40]]
              precision    recall  f1-score   support

          NO       0.80      0.85      0.82        66
         YES       0.80      0.74      0.77        54

    accuracy                           0.80       120
   macro avg       0.80      0.79      0.80       120
weighted avg       0.80      0.80      0.80       120

[[49 17]
 [32 22]]
              precision    recall  f1-score   support

          NO       0.60      0.74      0.67        66
         YES       0.56      0.41      0.47        54

    accuracy                           0.59       120
   macro avg       0.58      0.57      0.57       120
weighted avg       0.59      0.59      0.58       120

Tree test accuracy: 0.8, Tree train accuracy: 1.0, NB test accuracy: 0.5916666666666667, NB train accuracy: 0.65625


## <h5> The above results show that overall the decision tree would be the best model to use, as it has the highest accuracy on the test set at .8, beating out KNN at .625 and Naive-bayes at .59. However, the tree has a 1.0 accuracy on the training set, showing us the model is overfit to the data and thus has high variance, and this may suggest that the model won't do as well as we would hope in future test sets. The naive-bayes model is let fit to the model, and has a high bias, we can be seen due to its relatively low accuracy on its own training set. This showed itself in the test accuracy, where it was much less accurate than the tree model. The naive-bayes might do better than expected when given a new test set compared to the tree. </h5>